CRI: Add security context for sandbox/container

feiskyer opened the pull request:
Part of #29478. This PR 
- adds security context for sandbox and fixes #33139
- encaps container security context to `SecurityContext` and adds missing features
- Note that capability is not fully accomplished in this PR because it is under discussion at  #33614.

cc/ @yujuhong @yifan-gu @Random-Liu @kubernetes/sig-node

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34811)

<!-- Reviewable:end -->

feiskyer commented:
@k8s-bot cri test this.

feiskyer commented:
Didn't find way for triggering k8s-bot to test security context, so just tested in my local setting. 

There are 3 failed tests remaining, all related with `volume SELinux relabeling`:

```
Summarizing 3 Failures:

[Fail] [k8s.io] Security Context [Feature:SecurityContext] [It] should support volume SELinux relabeling when using hostPID 
/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/security_context.go:180

[Fail] [k8s.io] Security Context [Feature:SecurityContext] [It] should support volume SELinux relabeling 
/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/security_context.go:235

[Fail] [k8s.io] Security Context [Feature:SecurityContext] [It] should support volume SELinux relabeling when using hostIPC 
/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/security_context.go:180

Ran 10 of 419 Specs in 156.744 seconds
FAIL! -- 7 Passed | 3 Failed | 0 Pending | 409 Skipped --- FAIL: TestE2E (156.76s)
```

@pmorie @yujuhong Could you help to figure out what I'm missing?

yifan-gu commented:
What distro are you using? Have you tried to launch a pod with selinux enabled by hand?

feiskyer commented:
> What distro are you using? Have you tried to launch a pod with selinux enabled by hand?

@yifan-gu It's CentOS 7. Already tried and finally figured out the problem.

@pmorie @yifan-gu @yujuhong Fixed all tests. PTAL.

mrunalp commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -127,16 +127,52 @@ message NamespaceOption {
+    // The configurations for the sandbox's namespaces.
+    // This will be used only if the PodSandbox uses namespace for isolation.
+    optional NamespaceOption namespace_options = 2;
+    // Optional SELinux context to be applied.
```
Should add a sentence about precedence here as well.

feiskyer replied:
ack. thanks.

mrunalp commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -360,26 +396,40 @@ message Capability {
-    optional Capability capabilities = 2;
+    optional Capability capabilities = 1;
+    // If set, run container in privileged mode.
+    // Processes in privileged containers are essentially equivalent to root on the host.
```
It isn't clear how this is translated in case of user namespaces. So, maybe leave out this sentence?

feiskyer replied:
ack

mrunalp commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -360,26 +396,40 @@ message Capability {
     optional SELinuxOption selinux_options = 3;
-    // User contains the user for the container process.
-    optional LinuxUser user = 4;
+    // The UID to run the entrypoint of the container process.
```
Maybe just call it the container process? entrypoint is too docker specific.

mrunalp commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -360,26 +396,40 @@ message Capability {
+    // May also be set in PodSecurityContext.  If set in both SecurityContext and
+    // PodSecurityContext, the value specified here takes precedence.
+    optional int64 run_as_user = 4;
+    // Indicates that the container must run as a non-root user.
```
Is the runtime supposed to pick a non root uid if this is set or is run_as_user also expected to be set when this is set?

feiskyer replied:
y, run_as_user is expected to set, but it may also be set from image metadata. that's why this requires runtime to validate it.

pmorie commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,200 @@
+	"strconv"
+
+	dockercontainer "github.com/docker/engine-api/types/container"
+	runtimeApi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
```
I believe `runtimeapi` would be idiomatic -- we do not use camel case elsewhere in the code in import statements

feiskyer replied:
@pmorie y, @Random-Liu has already opened an issue for this #35070, we can fix it in a separate PR.  Also here should be fixed.

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -127,16 +127,52 @@ message NamespaceOption {
+    // A list of groups applied to the first process run in each container, in addition
+    // to the container's primary GID.  If unspecified, no groups will be added to
+    // any container.
+    repeated int64 supplement_groups = 7;
```
Rechecked with this, I think `supplemental_groups` is better (keep same with api)

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -360,26 +396,40 @@ message Capability {
+    // If set, run container in privileged mode.
+    // Processes in privileged containers are essentially equivalent to root on the host.
+    optional bool privileged = 2;
     // Optional SELinux context to be applied.
```
ack

feiskyer commented:
@mrunalp @pmorie Addressing comments. PTAL.

yifan-gu commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -127,16 +127,54 @@ message NamespaceOption {
+    // The UID to run the entrypoint of the container process.
+    // Defaults to user specified in image metadata if unspecified.
+    // May also be set in PodSecurityContext.  If set in both SecurityContext and
+    // PodSecurityContext, the value specified here takes precedence.
```
Why pod-level overrides container-level for this field?

feiskyer replied:
@yifan-gu Sorry for placing wrong docs. will fix.

yifan-gu commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -127,16 +127,54 @@ message NamespaceOption {
     optional bool host_ipc = 3;
 }
 
+// PodSecurityContext holds pod-level security configuration that will be applied to a container.
```
A little confused about PodSecurityContext vs SecurityContext.
Does PodSecurityContext functions as a default value when SecurityContext is not provided (in which case, the container level one overrides the pod level one).
or does PodSecurityContext overrides the SecurityContext??

yifan-gu replied:
The comments indicates both can happen (e.g. for selinux options, container level overrides the pod level, for run_as_user, the pod one overrides the container one)

feiskyer replied:
@yifan-gu My mistake of placing wrong docs. PodSecurityContext functions as a default value when SecurityContext is not provided and SecurityContext will override PodSecurityContext if both are set.

pmorie replied:
@feiskyer is right.  You can check out the proposal for more info: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/pod-security-context.md

Of course, I realize now that I never moved that proposal to 'designs', which is what you're supposed to do when you implement a proposal :) So, it'll be in designs soon.

feiskyer commented:
@yifan-gu Addessed comments and rebased. PTAL.

rhatdan commented:
From an SELinux point of view PodSecurityContext should be default and container security context would override.

feiskyer commented:
@rhatdan Exactly.

pmorie commented:
Reviewing this now.

pmorie commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -133,16 +133,54 @@ message NamespaceOption {
+    // This will be used only if the PodSandbox uses namespace for isolation.
+    optional NamespaceOption namespace_options = 2;
+    // Optional SELinux context to be applied. If set in both SecurityContext
+	// and PodSecurityContext, the value specified in SecurityContext takes
```
nit, there's an indentation issue here

feiskyer replied:
ack

pmorie commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -133,16 +133,54 @@ message NamespaceOption {
+message PodSecurityContext {
+    // The configurations for the sandbox's namespaces.
+    // This will be used only if the PodSandbox uses namespace for isolation.
+    optional NamespaceOption namespace_options = 2;
```
why is the field called `thing_options` and the message is called `<thing>Option`?

feiskyer replied:
GRPC will generate `ThingOptions` based on `thing_options`, so document it as `ThingOptions` to not break godoc conventions.

pmorie commented:
@feiskyer Is it intentional that this PR removes the ability to supply a primary GID for a `LinuxContainerSandboxConfig`?

pmorie commented:
While I'm thinking about it, I think the commits should be re-cut before this goes in.  Consider doing:

1.  A commit for the API changes to .proto
2.  A commit for the changes to api.pb.go
3.  A commit for other code changes

pmorie commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -230,25 +230,46 @@ func (ds *dockerService) makeSandboxDockerConfig(c *runtimeApi.PodSandboxConfig,
 		// TODO: Check if this works with per-pod cgroups.
```
Nit, you could extract a function here called `applyLinuxSpecificOptions` and the whole method would read as a transaction script on a uniform level of abstraction.

feiskyer replied:
ack.

pmorie commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,200 @@
+}
+
+// modifyHostConfig applies security context config to dockercontainer.HostConfig.
+func modifyHostConfig(sc *runtimeapi.SecurityContext, podSc *runtimeapi.PodSecurityContext, hostConfig *dockercontainer.HostConfig, isSandbox bool) {
```
Same thing here re: argument ordering.

pmorie commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -198,64 +186,76 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerConfig(container *api.
+			HostPath:       &v.HostPath,
+			ContainerPath:  &v.ContainerPath,
+			Readonly:       &v.ReadOnly,
+			SelinuxRelabel: &v.SELinuxRelabel,
```
This is consistent with #33951, I approve.

pmorie commented:
Looks like a solid start, left several comments.

feiskyer commented:
> @feiskyer Is it intentional that this PR removes the ability to supply a primary GID for a LinuxContainerSandboxConfig?

@pmorie Do you mean the removed `LinuxUser` here? Y, I removed `LinuxUser` intentionally because user is already set at `SecurityContext` and also makes `SecurityContext` consistent with api.SecurityContext.

feiskyer commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,200 @@
+}
+
+// modifyContainerConfig applies security context config to dockercontainer.Config.
+func modifyContainerConfig(sc *runtimeapi.SecurityContext, podSc *runtimeapi.PodSecurityContext, config *dockercontainer.Config) {
```
ack.

feiskyer commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,200 @@
+// modifyHostConfig applies security context config to dockercontainer.HostConfig.
+func modifyHostConfig(sc *runtimeapi.SecurityContext, podSc *runtimeapi.PodSecurityContext, hostConfig *dockercontainer.HostConfig, isSandbox bool) {
+	// Apply supplemental groups
+	// TODO: We skip application of supplemental groups to the
```
ack.

feiskyer commented:
> While I'm thinking about it, I think the commits should be re-cut before this goes in. Consider doing:
> 
> A commit for the API changes to .proto
> A commit for the changes to api.pb.go
> A commit for other code changes

@pmorie Re-cut commits and addressed comments. PTAL.

feiskyer commented:
waiting for #33663

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -394,26 +432,41 @@ message Capability {
+    // If set, run container in privileged mode.
+    optional bool privileged = 2;
+    // Optional SELinux context to be applied. If set in both SecurityContext
+	// and PodSecurityContext, the value specified here takes precedence
```
indent is slightly off in this line and the next

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -133,16 +133,54 @@ message NamespaceOption {
+    optional NamespaceOption namespace_options = 2;
+    // Optional SELinux context to be applied. If set in both SecurityContext
+    // and PodSecurityContext, the value specified in SecurityContext takes
+    // precedence for the container.
```
Should kubelet or the runtime decide enforce this? I feel like kubelet should supply this in the `ContainerConfig` to make things simpler.

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -133,16 +133,54 @@ message NamespaceOption {
+    // 2. The setgid bit is set (new files created in the volume will be owned by FSGroup)
+    // 3. The permission bits are OR'd with rw-rw----
+    //
+    // If unset, the Kubelet will not modify the ownership and permissions of any volume.
```
Is Kubelet going to do this, or the runtime?

pmorie replied:
The volume plugins that support it do it during `Setup`

feiskyer replied:
ack, will merge it to supplemental_groups

yujuhong reviewed (commented):
My understanding is that in k8s api, we allow users to specify a base config for all containers in the pod through `PodSecurityContext`, and a separate per-container `SecurityContext` to override the base config if needed. However, for CRI, I think kubelet should be the one resolving the two configs and only send the correct per-container `SecurityContext` to the container. There is really no need to specify twice in the podsandbox config and in the containe config. Of course, options that are only applicable to the entire pod should remain in the `PodSecurityContext`.

yujuhong commented:
> My understanding is that in k8s api, we allow users to specify a base config for all containers in the pod through PodSecurityContext, and a separate per-container SecurityContext to override the base config if needed. However, for CRI, I think kubelet should be the one resolving the two configs and only send the correct per-container SecurityContext to the container. There is really no need to specify twice in the podsandbox config and in the containe config. Of course, options that are only applicable to the entire pod should remain in the PodSecurityContext.

Basically, CRI should be less ambiguous and only send the runtime *one* set of parameters.

feiskyer commented:
>  I think kubelet should be the one resolving the two configs and only send the correct per-container SecurityContext to the container.

Good catch. This could make the api more clear. Will move the logic to kuberuntime.

feiskyer commented:
Addressed comments and renamed `PodSecurityContext` to `SandboxSecurityContext`. so now

- SandboxSecurityContext is only for sandbox and
- SecurityContext is only for containers
- No override now in CRI.

@yujuhong @timstclair PTAL.

( BTW, the pr is still waiting for #33663. )

feiskyer commented:
@k8s-bot unit test this

feiskyer commented:
Rebased because #33663 has just been merged. @pmorie @yujuhong PTAL.

pmorie commented on pkg/kubelet/api/v1alpha1/runtime/api.pb.go:
```diff
@@ -458,23 +459,112 @@ func (m *NamespaceOption) GetHostIpc() bool {
+	// to be owned by the pod:
+	//
+	// 1. The owning GID will be the FSGroup
+	// 2. The setgid bit is set (new files created in the volume will be owned by FSGroup)
```
Note to self, make sure this is accurate after fix to setting setgid bit on files is merged.

pmorie commented on pkg/kubelet/api/v1alpha1/runtime/api.pb.go:
```diff
@@ -458,23 +459,112 @@ func (m *NamespaceOption) GetHostIpc() bool {
+	RunAsNonRoot *bool `protobuf:"varint,4,opt,name=run_as_non_root,json=runAsNonRoot" json:"run_as_non_root,omitempty"`
+	// If set, the root filesystem of the container is read-only.
+	ReadonlyRootfs *bool `protobuf:"varint,5,opt,name=readonly_rootfs,json=readonlyRootfs" json:"readonly_rootfs,omitempty"`
+	// A list of groups applied to the first process run in the sandbox, in addition
```
These apply to the first process in each container run in the sandbox, accurate?

feiskyer replied:
Here is only for sandbox. Containers are addressed at `SecurityContext.SupplementalGroups`

pmorie commented on pkg/kubelet/api/v1alpha1/runtime/api.pb.go:
```diff
@@ -458,23 +459,112 @@ func (m *NamespaceOption) GetHostIpc() bool {
+	// Optional SELinux context to be applied. Only for sandbox. Containers should
+	// confirm to SecurityContext.SelinuxOptions.
+	SelinuxOptions *SELinuxOption `protobuf:"bytes,2,opt,name=selinux_options,json=selinuxOptions" json:"selinux_options,omitempty"`
+	// The UID to run the entrypoint of the container process.
```
s/process/process as/

pmorie commented on pkg/kubelet/api/v1alpha1/runtime/api.pb.go:
```diff
@@ -458,23 +459,112 @@ func (m *NamespaceOption) GetHostIpc() bool {
+	NamespaceOptions *NamespaceOption `protobuf:"bytes,1,opt,name=namespace_options,json=namespaceOptions" json:"namespace_options,omitempty"`
+	// Optional SELinux context to be applied. Only for sandbox. Containers should
+	// confirm to SecurityContext.SelinuxOptions.
+	SelinuxOptions *SELinuxOption `protobuf:"bytes,2,opt,name=selinux_options,json=selinuxOptions" json:"selinux_options,omitempty"`
```
Golang naming conventions would have this named 'SELinuxOptions', as it is elsewhere in the API, with the json field name being 'seLinuxOptions'.

feiskyer replied:
It's hard to ensure the field name for both grpc and golang conventions. Some options here

- Option 1: Named `SELinuxOptions` with json `sELinuxOptions`
- Option 2: Named `SeLinuxOptions` with json `seLinuxOptions`
- Option 3: Named `SelinuxOptions ` with json `selinuxOptions`

Option 3 seems most comfortable to me. @pmorie What's your opinion?

timstclair replied:
I agree with @feiskyer . Since the proto is the source for this API, I think it should adhere to proto convensions, even when that means the generated code violates go conventions.

yujuhong replied:
Generated code are exempted from adhering to the golang convention: https://github.com/golang/go/wiki/CodeReviewComments#initialisms

pmorie reviewed (commented):
Didn't get all the way through, will finish on plane

yujuhong commented:
@pmorie, I think kubelet should leave as little room for runtime to interpret the security context as possible, so runtime should not be overlaying contexts for containers. 
It seems like we can categorize the settings in 3 sections.
  1. In the podsandbox config, define options that should be applied to all containers (e.g., host network); no override at the container level is allowed.
  2. In the podsandbox config, define options that *may be* applicable to the sandbox itself. Since sandbox is interpreted by each runtime differently, and may or may not involve a running process, this section may be meaningless for some runtimes.
  3. In the container config, define options that should be applied to this particular container. Options in (3) may not overlapped with options in (1). 

This PR implements (2) and (3), and folds (1) into (3) for simplicity. Does that make sense to you?

timstclair commented on pkg/kubelet/dockershim/security_context_test.go:
```diff
@@ -0,0 +1,310 @@
+	for _, tc := range cases {
+		dockerCfg := &dockercontainer.Config{}
+		modifyContainerConfig(tc.sc, dockerCfg)
+		if e, a := tc.expected, dockerCfg; !reflect.DeepEqual(e, a) {
```
Use `assert.Equal`

timstclair commented on pkg/kubelet/dockershim/security_context_test.go:
```diff
@@ -0,0 +1,310 @@
+		dockerCfg := &dockercontainer.HostConfig{}
+		modifyHostConfig(tc.sc, "", dockerCfg)
+
+		if e, a := tc.expected, dockerCfg; !reflect.DeepEqual(e, a) {
```
`assert.Equal` - same below.

feiskyer replied:
ack

timstclair commented on pkg/kubelet/dockershim/legacy.go:
```diff
@@ -49,3 +49,7 @@ func (ds *dockerService) LegacyPortForward(sandboxID string, port uint16, stream
 }
+
+func (ds *dockerService) isImageRoot(image string) (bool, error) {
+	return dockertools.IsImageRoot(ds.client, image)
```
nit: just inline this.

feiskyer replied:
ack

timstclair commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -127,26 +127,47 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *api.Pod, attem
+				HostPid:     &sc.HostPID,
+			},
+		}
+		if sc.RunAsUser != nil {
```
nit: you don't need to check this. Same below.

feiskyer replied:
ack

timstclair commented on pkg/kubelet/kuberuntime/security_context.go:
```diff
@@ -0,0 +1,150 @@
+)
+
+// determineEffectiveSecurityContext gets container's security context from api.Pod and container.SecurityContext.
+func (m *kubeGenericRuntimeManager) determineEffectiveSecurityContext(pod *api.Pod, sc *api.SecurityContext) *runtimeapi.SecurityContext {
```
This looks like a dupe of https://github.com/kubernetes/kubernetes/blob/master/pkg/securitycontext/provider.go#L125? Why not use that?

feiskyer replied:
Yes, most of it is stolen from `pkg/securitycontext`. But the package is very docker specific, so I moved the logic to dockershim/kuberuntime. we should remove the package after docker is fully moved to dockershim.

timstclair replied:
The method isn't docker specific, and is already called from the rkt package. I agree that package should be refactored a bit to make it less docker-centric, but I don't think that's a good reason to duplicate code.

feiskyer replied:
Another reason is `determineEffectiveSecurityContext ` returns `runtimeapi.SecurityContext` instead of `api.SecurityContext`, and `runtimeapi.SecurityContext` contains more fields than `api.SecurityContext`. 

Will change to `DetermineEffectiveSecurityContext()`.

timstclair commented on pkg/kubelet/kuberuntime/security_context.go:
```diff
@@ -0,0 +1,150 @@
+	if podSc.SELinuxOptions != nil {
+		synthesized.SelinuxOptions = convertToRuntimeSELinuxOption(podSc.SELinuxOptions)
+	}
+	if podSc.RunAsUser != nil {
```
ditto: don't need to check for nil

feiskyer replied:
ack

timstclair commented on pkg/kubelet/kuberuntime/security_context.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+// securityContextFromPodSecurityContext gets security context from api.Pod.
+func (m *kubeGenericRuntimeManager) securityContextFromPodSecurityContext(pod *api.Pod) *runtimeapi.SecurityContext {
```
You don't need this method - just resolve the pod security context to a container security context, and then use the same conversion.

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -133,16 +133,50 @@ message NamespaceOption {
+    // Only for sandbox. Containers should confirm to SecurityContext.RunAsNonRoot.
+    optional bool run_as_non_root = 4;
+    // If set, the root filesystem of the container is read-only.
+    optional bool readonly_rootfs = 5;
```
I'm not sure whether it is runtime-specific or not. Docker sets this for sandbox today; other runtimes may also want to perform safe behavior with docker. Since ReadOnlyRootfs is part of api.PodSecurityContext, I think CRI should expose it to runtimes and lets runtime to decide how to deal with it.

@yujuhong What do you think?

timstclair replied:
I just mean for the sandbox it seems runtime specific. I'm happy to keep it on the container. Since the user has no (direct) control over what the sandbox is doing, it seems like they would always want to set `readonly_rootfs` to true, but it's not clear what the consequences would be. The only counter argument I can think of is if `readonly_rootfs` on the sandbox depends on whether all the containers have a `readonly_rootfs`. This would still be pretty runtime specific though, and might warrant a larger change to the API.

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -133,16 +133,50 @@ message NamespaceOption {
+    // If set, the root filesystem of the container is read-only.
+    optional bool readonly_rootfs = 5;
+    // A list of groups applied to the first process run in the sandbox, in addition
+    // to the container's primary GID.  If unspecified, no groups will be added.
```
ack

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -394,26 +428,49 @@ message Capability {
+    // If true, the runtime should validate the image at runtime to ensure that it
+    // does not run as UID 0 (root) and fail to start the container if it does.
+    // If unset or false, no such validation will be performed.
+    optional bool run_as_non_root = 6;
```
They are in different scopes:

Processes in privileged containers are essentially equivalent to root on the host; while containers run with root user are still be constrained, e.g. runtimes may map the root user of container to an ordinary user on the host.

yujuhong replied:
Can we move this check to kubelet? Kubelet can read the image metdata, verify `run_as_non_root`, and always set  `run_as_user` when running a container. Then we can remove `run_as_non_root` here.

feiskyer replied:
yes, but there is no user in current image metadata, needs extend it first.

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -394,26 +428,49 @@ message Capability {
+    // 3. The permission bits are OR'd with rw-rw----
+    //
+    // If unset, the Kubelet will not modify the ownership and permissions of any volume.
+    optional int64 fs_group = 9;
```
Yes, but runtimes may need this while creating containers, e.g. in docker, it is set in `hostConfig.GroupAdd`.

timstclair replied:
I see. So we expect the Kubelet to pass the same values for both then?

yujuhong replied:
I don't think we need both `supplemental_groups` and `fs_groups` since they essentially mean the same thing for the runtime. Can we merge them? The same applies to the sandbox security context.

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -473,19 +526,18 @@ message ContainerConfig {
 
     // Linux contains configuration specific to Linux containers.
-    optional LinuxContainerConfig linux = 16;
+    optional LinuxContainerConfig linux = 14;
```
exactly, once released, these numbers shouldn't be changed. we can arrange these fields before release to make it looking comfortable.

yujuhong replied:
Yes, we should not be arbitrarily changing them (or it would be breaking) once we release API. Now we assume developers know that they are developing against a pre-alpha api.

feiskyer commented on pkg/kubelet/dockershim/docker_container.go:
```diff
@@ -174,13 +155,32 @@ func (ds *dockerService) CreateContainer(podSandboxID string, config *runtimeApi
 		// Note: ShmSize is handled in kube_docker_client.go
+
+		// Verify RunAsNonRoot.
+		if sc := lc.GetSecurityContext(); sc != nil && sc.RunAsNonRoot != nil {
```
ack

feiskyer commented on pkg/kubelet/dockershim/docker_container.go:
```diff
@@ -174,13 +155,32 @@ func (ds *dockerService) CreateContainer(podSandboxID string, config *runtimeApi
+		// Verify RunAsNonRoot.
+		if sc := lc.GetSecurityContext(); sc != nil && sc.RunAsNonRoot != nil {
+			if sc.RunAsUser != nil {
+				return "", fmt.Errorf("container's runAsUser breaks non-root policy")
```
Hmm, my mistake, should be `if sc.GetRunAsUser() == 0`

feiskyer commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -203,6 +203,33 @@ func (ds *dockerService) ListPodSandbox(filter *runtimeApi.PodSandboxFilter) ([]
 }
 
+// applyLinuxSpecificOptions applies LinuxPodSandboxConfig to dockercontainer.HostConfig and dockercontainer.ContainerCreateConfig.
+func (ds *dockerService) applyLinuxSpecificOptions(hc *dockercontainer.HostConfig, lc *runtimeApi.LinuxPodSandboxConfig, createConfig *dockertypes.ContainerCreateConfig, image string) error {
```
ack

feiskyer commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -203,6 +203,33 @@ func (ds *dockerService) ListPodSandbox(filter *runtimeApi.PodSandboxFilter) ([]
+	// TODO: Check if this works with per-pod cgroups.
+	hc.CgroupParent = lc.GetCgroupParent()
+
+	// Verify RunAsNonRoot of security context.
```
For simplicity

- The verification is related with docker client ( it calls `ds.isImageRoot` to query image user), while applySandboxSecurityContext doesn't require the client.
- applySandboxSecurityContext never error out

feiskyer commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -203,6 +203,33 @@ func (ds *dockerService) ListPodSandbox(filter *runtimeApi.PodSandboxFilter) ([]
+	hc.CgroupParent = lc.GetCgroupParent()
+
+	// Verify RunAsNonRoot of security context.
+	if podSc := lc.GetSecurityContext(); podSc != nil && podSc.RunAsNonRoot != nil {
```
ack

feiskyer commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,199 @@
+		hostConfig.SecurityOpt = modifySecurityOption(hostConfig.SecurityOpt, dockerLabelUser, sc.SelinuxOptions.GetUser())
+		hostConfig.SecurityOpt = modifySecurityOption(hostConfig.SecurityOpt, dockerLabelRole, sc.SelinuxOptions.GetRole())
+		hostConfig.SecurityOpt = modifySecurityOption(hostConfig.SecurityOpt, dockerLabelType, sc.SelinuxOptions.GetType())
+		hostConfig.SecurityOpt = modifySecurityOption(hostConfig.SecurityOpt, dockerLabelLevel, sc.SelinuxOptions.GetLevel())
```
Yes, it's from `pkg/securitycontext/provider.go`

timstclair replied:
Mentioned below, we should reuse that logic.

feiskyer commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,199 @@
+// of name:value
+func modifySecurityOption(config []string, name, value string) []string {
+	if len(value) > 0 {
+		config = append(config, fmt.Sprintf("%s:%s", name, value))
```
Here is SELinux options, not docker's options. Maybe should rename it more clearly.

timstclair replied:
1. I'm pretty sure that logic is wrong too. I doubt Kubernetes SELinux works with docker 1.11+ right now (/cc @pmorie )
2. We should reuse the logic you linked to rather than duplicating it.

feiskyer commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,199 @@
+	if podSc != nil && podSc.RunAsUser != nil {
+		modifyContainerConfig(&runtimeapi.SecurityContext{
+			RunAsUser: podSc.RunAsUser,
+		}, config)
```
Here casts `SandboxSecurityContext` to `SecurityContext`. It could reuse `modifyContainerConfig` acutally. Will update.

timstclair replied:
I still think the internal type might be less confusing, but I'm ok with using the container SecurityContext for that purpose. However, I think you can just adapt the `SandboxSecurityContext` to a `SecurityContext` once in `applySandboxSecurityContext`, and then you don't need 2 versions of `modify*ContainerConfig` and `modify*HostConfig`

feiskyer commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,199 @@
+}
+
+// modifyNamespaceOptions applies namespaceoptions to dockercontainer.HostConfig.
+func modifyNamespaceOptions(sc *runtimeapi.SecurityContext, sandboxID string, hostConfig *dockercontainer.HostConfig) {
```
ack

feiskyer commented:
@timstclair Addressed comments and rebased. PTAL.

timstclair commented on pkg/kubelet/dockershim/security_context_test.go:
```diff
@@ -0,0 +1,310 @@
+	bothHC.GroupAdd = []string{"2222", "1234"}
+	fsGroup := int64(1234)
+
+	testCases := map[string]struct {
```
nit: In all the other tests you use a name field in a struct for the testcase name, but here you use a map. Stick with one for consistency.

feiskyer replied:
ack

timstclair commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -64,7 +65,7 @@ func modifySandboxContainerConfig(podSc *runtimeapi.SandboxSecurityContext, conf
 func modifyContainerConfig(sc *runtimeapi.SecurityContext, config *dockercontainer.Config) {
 	if sc != nil && sc.RunAsUser != nil {
-		config.User = strconv.Itoa(int(*sc.RunAsUser))
+		config.User = strconv.Itoa(int(sc.GetRunAsUser()))
```
Use `strconv.FormatInt` so you don't lose precision.

timstclair commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,200 @@
+// modifyContainerConfig applies container security context config to dockercontainer.Config.
+func modifyContainerConfig(sc *runtimeapi.SecurityContext, config *dockercontainer.Config) {
+	if sc != nil && sc.RunAsUser != nil {
+		config.User = strconv.Itoa(int(sc.GetRunAsUser()))
```
Use `strconv.FormatInt` so you don't lose precision.

feiskyer replied:
ack

timstclair commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -394,26 +428,49 @@ message Capability {
 }
 
+// SecurityContext holds security configuration that will be applied to a container.
+message SecurityContext {
```
Can we name this `ContainerSecurityContext`? I think the only reason it's not that way in the k8s API was that it was added before there was a `PodSecurityContext`

feiskyer replied:
ack

timstclair commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -133,16 +133,50 @@ message NamespaceOption {
-    // This will be used only if the PodSandbox uses namespace for isolation.
-    optional NamespaceOption namespace_options = 2;
+    // SandboxSecurityContext holds sandbox security attributes.
+    optional SandboxSecurityContext security_context = 2;
```
If we think about eventually incorporating other OSs, I think we should decide whether each would have a separate SecurityContext message (in which case these should be called `Linux{Sandbox,Container}SecurityContext`, or combine the options in the same message (similar to [docker HostConfig](https://godoc.org/github.com/docker/docker/api/types/container#HostConfig)) and move the SecurityContext out of the Linux config message.

feiskyer replied:
Current security context is very linux-specific, it's ok rename to `Linux{Sandbox,Container}SecurityContext`.

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.pb.go:
```diff
@@ -458,23 +459,112 @@ func (m *NamespaceOption) GetHostIpc() bool {
 	return false
 }
 
+// SandboxSecurityContext holds security configuration that will be applied to a sandbox.
```
ack

feiskyer commented on pkg/kubelet/api/v1alpha1/runtime/api.pb.go:
```diff
@@ -458,23 +459,112 @@ func (m *NamespaceOption) GetHostIpc() bool {
+	// This will be used only if the PodSandbox uses namespace for isolation.
+	NamespaceOptions *NamespaceOption `protobuf:"bytes,1,opt,name=namespace_options,json=namespaceOptions" json:"namespace_options,omitempty"`
+	// Optional SELinux context to be applied. Only for sandbox. Containers should
+	// confirm to SecurityContext.SelinuxOptions.
```
ack

feiskyer commented:
@timstclair  @yujuhong Addressed comments. PTAL.

feiskyer commented:
> I'm pretty sure that logic is wrong too. I doubt Kubernetes SELinux works with docker 1.11+ right now (/cc @pmorie )

@timstclair You are right about this. From docker's [documentation](https://docs.docker.com/engine/reference/run/#/security-configuration), it has been changed to

```
--security-opt="label=user:USER"   : Set the label user for the container
--security-opt="label=role:ROLE"   : Set the label role for the container
--security-opt="label=type:TYPE"   : Set the label type for the container
--security-opt="label=level:LEVEL" : Set the label level for the container
--security-opt="label=disable"     : Turn off label confinement for the container
```

so we need to change `label:` to `label=` for docker 1.11+. I'd like to fix the issue in the following PR.  cc/ @pmorie 

By the way, this PR has passed all Feature:SecurityContext tests in my local setting with Docker v1.12 (Maybe there is something missed in feature test?):

```
SSSSSSSSSSSS
Ran 10 of 428 Specs in 200.440 seconds
SUCCESS! -- 10 Passed | 0 Failed | 0 Pending | 418 Skipped PASS

$ docker --version
Docker version 1.12.3, build 6b644ec
```

@pmorie 's original PR (#33663) has also passed full Feature:SecurityContext tests (not sure which docker version).

yujuhong commented on pkg/kubelet/api/v1alpha1/runtime/api.proto:
```diff
@@ -707,6 +729,8 @@ message Image {
     // The size of the image in bytes.
     optional uint64 size = 4;
+    // The uid that will run the command(s).
+    optional int64 uid = 5;
```
@euank @yifan-gu are you ok with this or do you prefer runtime itself checks the `run_as_non_root` option? 

Background: In the initial version, kubelet set the `run_as_non_root` field in the security context and pass it to the runtime (via CRI). Runtime checks the (docker) image and verify that the user is not set to root. This assumes that the `RuntimeService` is aware of the images and can perform checks. Since we may want to decouple `ImageService` and `RuntimeService` in the future, I suggested move up this logic and let kubelet perform the check. This requires reporting the user name or ID as part of the image for kubelet to check.

euank replied:
Yeah, we have that image info handy as well.

I'm :+1: for kubelet doing that check.

My nit for this field is that `user` probably makes more sense since it can be either a name or uid, and the comment needs to clarify that dual-meaning as well.

euank replied:
Actually, reading more it looks like this is strictly only UID and k8s just doesn't support `non_root_user` for images that have a non-numeric user (hey, another Dockerfile feature that's not fully supported :).

looks like my concern isn't valid.

feiskyer committed:
CRI: add security context for sandbox/container

feiskyer committed:
run hack/update-generated-runtime.sh

feiskyer committed:
run hack/update-bazel.sh

feiskyer committed:
Implement security context in kuberuntime

feiskyer committed:
Add security context support in dockershim

feiskyer commented:
@timstclair @yujuhong  Rebased and fixed verification problems. PTAL.

timstclair commented:
LGTM

euank commented:
LGTM

pmorie commented:
I'd like a chance to re-review this before it merges, if it pleases the
court

On Friday, November 4, 2016, Yu-Ju Hong <notifications@github.com> wrote:

> *@yujuhong* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34811#pullrequestreview-7287216>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWXmAA3MEMqV1YfzKGvrFFZ2K0l6zRwks5q66J1gaJpZM4KW5AD>
> .
>

timstclair commented:
@pmorie - go ahead if you can review this weekend (I removed the LGTM). Let's not hold this up on minor details, we can iterate on those. If you don't have any major concerns, please re-apply the LGTM.

pmorie commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,158 @@
+	if sandboxID == "" {
+		modifyHostNetworkOptionForSandbox(hostNetwork, hostConfig)
+	} else {
+		// Set for container is sandboxID is provided.
```
s/container is/container if/

feiskyer replied:
ack

pmorie commented on pkg/kubelet/dockershim/security_context_test.go:
```diff
@@ -0,0 +1,257 @@
+				SelinuxOptions: inputSELinuxOptions(),
+			},
+			expected: setSELinuxHC,
+		},
```
How about a non-nil but empty security context?

feiskyer replied:
ack

pmorie commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -300,9 +272,11 @@ func (m *kubeGenericRuntimeManager) makeMounts(opts *kubecontainer.RunContainerO
 			glog.Errorf("Error on creating termination-log file %q: %v", containerLogPath, err)
 		} else {
 			fs.Close()
+			selinuxRelabel := selinux.SELinuxEnabled()
```
nit, i am not sure what this var buys us here

feiskyer replied:
@pmorie This is for setting its pointer in following lines.

pmorie commented on pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:
```diff
@@ -127,26 +127,43 @@ func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *api.Pod, attem
 
-	if cgroupParent != "" {
-		linuxPodSandboxConfig.CgroupParent = &cgroupParent
+		if groups := m.runtimeHelper.GetExtraSupplementalGroupsForPod(pod); len(groups) > 0 {
```
Missed FSGroup here, if I'm not mistaken.

feiskyer replied:
ack

feiskyer commented on pkg/kubelet/dockershim/helpers.go:
```diff
@@ -123,13 +123,18 @@ func extractLabels(input map[string]string) (map[string]string, map[string]strin
+		// to the volume to the container which mounts it first.
+		// TODO: always relabel if SELinux is enabled and the volume support relabeling
+		// (refer #33951 and #33663).
 		if m.GetSelinuxRelabel() {
```
@pmorie Already handled this at [here](https://github.com/kubernetes/kubernetes/pull/34811/files#diff-44f60dd6d99cb695e5f333647ebd0703R250), here's todo should be removed.

feiskyer commented on pkg/kubelet/dockershim/security_context.go:
```diff
@@ -0,0 +1,158 @@
+	var sc *runtimeapi.LinuxContainerSecurityContext
+	if lc.SecurityContext != nil {
+		sc = &runtimeapi.LinuxContainerSecurityContext{
+			// TODO: We skip application of supplemental groups to the
```
ack

feiskyer committed:
cleanup TODOs in dockershim

feiskyer commented:
@pmorie Rebased and addressed comments. PTAL.

yujuhong commented:
@pmorie does this looks good to you?

yujuhong commented:
I think most reviewers have went through this PR a couple of times, and agreed with the overall direction. I'll leave the PR open for one more hour for reviewers to leave feedback. Will apply lgtm at 3PM PST if no one responds.
